internal/fetchers/oci/files/files.go

```diff
@@ -0,0 +1,89 @@
+// Copyright The Enterprise Contract Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package files
+
+import (
+	"archive/tar"
+	"context"
+	"encoding/json"
+	"io"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+
+	"github.com/enterprise-contract/ec-cli/internal/fetchers/oci"
+)
+
+type extractor interface {
+	fetcherFor(v1.Image) (fetcher, error)
+}
+
+type fetcher interface {
+	fetch(*tar.Header, io.Reader) (json.RawMessage, error)
+}
+
+var supported = []extractor{
+	olmManifestExtractor{},
+}
+
+func ImageFiles(ctx context.Context, ref name.Reference, opts ...remote.Option) (map[string]json.RawMessage, error) {
+	img, err := oci.NewClient(ctx).Image(ref, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	fetchers := make([]fetcher, 0, len(supported))
+	for _, f := range supported {
+		if f, err := f.fetcherFor(img); err != nil {
+			return nil, err
+		} else if f != nil {
+			fetchers = append(fetchers, f)
+		}
+	}
+
+	if len(fetchers) == 0 {
+		return nil, err
+	}
+
+	content := mutate.Extract(img)
+	defer content.Close()
+	archive := tar.NewReader(content)
+
+	manifests := map[string]json.RawMessage{}
+	for {
+		header, err := archive.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+
+		for _, f := range fetchers {
+			if data, err := f.fetch(header, archive); err != nil {
+				return nil, err
+			} else if data != nil {
+				manifests[header.Name] = data
+				break
```

Comment:
So we only care about the first successful fetcher? That seems odd. What if two fetchers want to fetch the same file? If a fetcher can manipulate the file it is fetching, `olmManifestExtractor` does, this will be a problem.

If we want to have multiple fetchers, we should probably namespace their result, e.g. `manifests.olm["manifests/crd.yaml"]`, where `olm` is the name of the fetcher.

Reply:
Yup, we either have all the files being fetched the same way, or we need to namespace them. Either approach is fine IMO.

(We could also make the fetcher just return a list of files to fetch.)